wfe2/verify.go

```diff
@@ -506,3 +549,85 @@ func (wfe *WebFrontEndImpl) validSelfAuthenticatedPOST(
 
 	return payload, pubKey, nil
 }
+
+// rolloverRequest is a struct representing an ACME key rollover request
+type rolloverRequest struct {
+	NewKey  jose.JSONWebKey
+	Account string
+}
+
+// validKeyRollover checks if the innerJWS is a valid key rollover operation
+// given the outer JWS that carried it. It is assumed that the outerJWS has
+// already been validated per the normal ACME process using `validPOSTForAccount`.
+// It is *critical* this is the case since `validKeyRollover` does not check the
+// outerJWS signature. This function checks that:
+// 1) the inner JWS is valid and well formed
+// 2) the inner JWS has the same "url" header as the outer JWS
+// 3) the inner JWS is self-authenticated with an embedded JWK
+// 4) the payload of the inner JWS is a valid JSON key change object
+// 5) that the key specified in the key change object *also* verifies the inner JWS
+// A *rolloverRequest object and is returned if successfully validated,
+// otherwise a problem is returned. The caller is left to verify whether the
+// new key is appropriate (e.g. isn't being used by another existing account)
+// and that the account field of the rollover object matches the account that
+// verified the outer JWS.
+func (wfe *WebFrontEndImpl) validKeyRollover(
+	outerJWS *jose.JSONWebSignature,
+	innerJWS *jose.JSONWebSignature,
+	logEvent *requestEvent) (*rolloverRequest, *probs.ProblemDetails) {
+
+	// Extract the embedded JWK from the inner JWS
+	jwk, prob := wfe.extractJWK(innerJWS)
+	if prob != nil {
+		return nil, prob
+	}
+
+	// If the key doesn't meet the GoodKey policy return a problem immediately
+	if err := wfe.keyPolicy.GoodKey(jwk.Key); err != nil {
+		wfe.stats.Inc("Errors.InnerJWKRejectedByGoodKey", 1)
+		return nil, probs.Malformed(err.Error())
+	}
+
+	// Check that the public key and JWS algorithms match expected
+	if err := checkAlgorithm(jwk, innerJWS); err != nil {
+		return nil, probs.Malformed(err.Error())
+	}
+
+	// Verify the inner JWS signature with the public key from the embedded JWK.
+	// NOTE(@cpu): We do not use `wfe.validJWSForKey` here because the inner JWS
+	// of a key rollover operation is special (e.g. has no nonce, doesn't have an
+	// HTTP request to match the URL to)
+	innerPayload, err := innerJWS.Verify(jwk)
+	if err != nil {
+		wfe.stats.Inc("Errors.InnerJWSVerificationFailed", 1)
+		return nil, probs.Malformed("Inner JWS does not verify with embedded JWK")
+	}
+	// NOTE(@cpu): we do not stomp the requestEvent's payload here since that is set
+	// from the outerJWS in validPOSTForAccount and contains the inner JWS and inner
+	// payload already.
+
+	// Verify that the outer and inner JWS protected URL headers match
+	prob = matchJWSURLs(outerJWS, innerJWS)
```

Comment:
LOL. I think you flagged the same thing on the original V2 JWS work and I forgot to do a pass this time to catch it myself :sweat:  Sorry about that